Statistics

bananabr3d opened the pull request:
TTTK-55: Manages statistics.db. Can add, update and delete statistics. Identification of statistics via player uuid.

bananabr3d committed:
feat: profile_save.py and profiles.json added, saves profiles and returns saved profiles

bananabr3d committed:
Merge remote-tracking branch 'origin/network_client' into network_client

bananabr3d committed:
Merge remote-tracking branch 'origin/network_client' into network_client

bananabr3d committed:
feat: TTTK-55 statistics.py added: provides different getters and add/update-methods. Also can count emojis and saves them into db.

bananabr3d committed:
fix: TTTK-55 requirements.txt: emoji lib ver 2.10.1 added to the requirements.txt

bananabr3d committed:
fix: TTTK-55 statistics.db: file added

bananabr3d committed:
feat(statistics): delete_all_statistics() added
deletes all records in the table statistics
[TTTK-55]

bananabr3d committed:
fix(statistics): getters raise ValueError if profile doesn't exist
instead of returning None
delete_statistics(uuid:str) and reset_statistics check for existance of stats, otherwise throws ValueError
[TTTK-55]

bananabr3d committed:
fix(statistics): commented one test loc
[TTTK-55]

bananabr3d committed:
chore(statistics): todos deleted
[TTTK-55]

bananabr3d committed:
fix(statistics): get_moves() raises error if uuid not found
[TTTK-55]

bananabr3d committed:
test(statistics): test_statistics.py: 13 tests for statistics.py
statistics.py test lines deleted
[TTTK-55]

bananabr3d committed:
test(statistics): abs path fixed

bananabr3d commented:
tests have been added and fixed :D

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+        :return: True if the file exists, False if it does not
+        """
+        if exists(path):
+            print("found")
```
use logging, not print

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+            print("found")
+            return True
+        else:
+            print("not found")
```
use logging, not print

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+                        if profile["profile_uuid"] == profile_uuid:
+                            return profile
+            except:
+                print("json error: Make sure profiles.json is formatted correctly")
```
use logging, not print

Petzys commented on json_schema/profiles.json:
Falls das die Datei sein soll, wo der Nutzer sein Profil persistiert, ist das nicht die richtige File Location. Eher unter /Client

Petzys replied:
Das ist ja gar kein JSON Schema

Petzys commented on Server/test_statistics.py:
```diff
@@ -0,0 +1,148 @@
+            self.statistics.get_wins("test")
+
+
+test = TestStatistics()
```
Bitte mit `__name__ = __main__`. Nicht so.

bananabr3d committed:
fix(statistics): winrate: try except added for case if 0 games played returns 0

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+                    with open(path, 'w') as file:
+                        json.dump(data, file)
+            except:
+                print("json error: Make sure profiles.json is formatted correctly")
```
use logging, not print

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+                        if profile["profile_name"] == profile_name:
+                            return profile
+            except:
+                print("json error: Make sure profiles.json is formatted correctly")
```
use logging, not print

Petzys replied:
Sollte diese Funktion nicht tatsächlich einen Error werfen, wenn das JSON halt falsch ist? Also bringt ja nichts, wenn das hier geloggt wird, wenn hier ein Fehler vorliegt.

Petzys commented on Client/profile_save.py:
```diff
@@ -0,0 +1,120 @@
+                entry = [{"profile_name": profile_name, "profile_uuid": profile_uuid, "profile_color": profile_color}]
+                json.dump(entry, file)
+
+#Testing
+#profile = Profile()
+#profile.add_new_profile("test", "test", "test")
+#print(profile.get_profiles())
+#print(profile.get_profile("test"))
+#profile.set_profile("test", "test2", "test3")
+#print(profile.get_profiles())
```
Remove

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+            with self.conn:
+                self.cursor.execute(f"""
+                DELETE FROM statistics
+                WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+                self.cursor.execute(f"""
+                UPDATE statistics
+                SET {arg} = 0
+                WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+                self.cursor.execute(f"""
+                UPDATE statistics
+                SET {type} = {type} + 1
+                WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+                self.cursor.execute(f"""
+                UPDATE statistics
+                SET moves = moves + {moves}
+                WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+                self.cursor.execute(f"""
+                UPDATE statistics
+                SET emojis = emojis + :emoji
+                WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+        with self.conn:
+            self.cursor.execute(f"""
+            SELECT * FROM statistics
+            WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,408 @@
+        with self.conn:
+            self.cursor.execute(f"""
+            SELECT * FROM statistics
+            WHERE uuid = '{uuid}'
```
This is a SQL Injection. Maybe use prepared Statements.

Petzys commented on Client/profile_save.py:
In diesem ganzen Modul solltest du immer mit den echten Profiles aus Server/Profiles.py arbeiten. Du solltest Profile Objects und keine Dictionaries zurückgeben. Das macht es leichter.

bananabr3d committed:
fix(statistics): sql injections fixed

JM-Lemmi commented on Client/profile_save.py:
why is this here?

JM-Lemmi commented on Server/Data/statistics.db:
gitignore!

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+    """
+    ________Public Methods_________
+    ________Main Methods___________
+    """
```
?

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+        Returns all statistics from the database
+        :return: an array of tuples containing:
+        (uuid : str, wins_first : int, wins_second : int, loses_first : int,
+        loses_second : int, draws : int, moves : int, emojis : int)
```
das ist ja wohl ein dict

Petzys commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+                          draws_first = draws_first, draws_second = draws_second,
+                          moves = moves, emojis = emojis)
+
+    def update_statistics(self, uuid: str, type: str, moves: int) -> None:
```
update_statistics should call add_profile if the profile does not exist. This makes this a lot easier for the server

JM-Lemmi replied:
+1

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+                          draws_first = draws_first, draws_second = draws_second,
+                          moves = moves, emojis = emojis)
+
+    def update_statistics(self, uuid: str, type: str, moves: int) -> None:
```
this call always needs a get call beforehand. please add an "increment_moves" function or similar.

JM-Lemmi replied:
No it doesnt, it does actually increment, but the docstring is wrong

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+        :param uuid: uuid of the profile that is updated
+        :param message: message that is checked for emojis
+        """
+        self._update_emojis(uuid, emoji.emoji_count(message))
```
this doesnt increment but overwrite the emoji count

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+        :param message: message that is checked for emojis
+        """
+        self._update_emojis(uuid, emoji.emoji_count(message))
+
```
missing function to increment winrate

JM-Lemmi commented on Server/statistics.py:
```diff
@@ -1,3 +1,409 @@
+                    ) -> None:
+        """
+        Adds a new profile to the database
+        :param uuid: uuid of the profile that is added
```
missing profile name

Petzys committed:
Feat(statistics): TTTK-55 Added JSON Schema for Statistics

JM-Lemmi commented:
functions I need as the server:

```
increment_moves(player, count = 1)
increment_emoji(player, message)
increment_wins_fromstate(gamestate)
```

Petzys committed:
Feat(statistics): TTTK-55 Added Client handling for statistics

bananabr3d committed:
chore(statistics): gitignore *.db added
feat: statistics.py reworked

Petzys committed:
Fix(statistics): TTTK-55 Fix JSON schema and client code to fit new statistics

JM-Lemmi committed:
refactor(statistics): rework statistics
only track wins, losses, draws, regardless of beginning player
use player objects
ref _check_add_profile() method
remove option to add profile with initial values. all values start at 0
[TTTK-55]

JM-Lemmi committed:
tests(statistics): rewrite tests for new refactored class
small bug fixes

JM-Lemmi committed:
Merge branch 'main' into statistics

JM-Lemmi committed:
feat(server): track statistics
[TTTK-68]

JM-Lemmi committed:
feat(server): send statistics on lobby join
[TTTK-69]

Petzys reviewed (approved):
Nice work!

Petzys committed:
Fix(ui client): Made statistics requestable

Petzys committed:
Fix(statistics): TTTK-70 Fixes incorrect dict access

Petzys committed:
Fix(server): TTTK-68 Fix incorrect parameter for statistics function, fix incorrect await

Petzys commented:
After testing I fixed some minor bugs.
